exercises/practice/perceptron/.docs/instructions.md

```diff
@@ -0,0 +1,34 @@
+# Instructions
+
+### Introduction
+[Perceptron](https://en.wikipedia.org/wiki/Perceptron) is one of the oldest and bestestly named machine learning algorithms out there. Since perceptron is also quite simple to implement, it's a favorite place to start a machine learning journey. As a linear classifier, if a linear decision boundary (e.g. a line in 2D or hyperplane in general) can be drawn to separate two labled classes of objects, perceptron is guaranteed to do so. This can help in predicting what an unlabeled object would likely be classified as by seeing which side of the decision boundary it is on.
+
+### Details
+The basic idea is fairly straightforward. We cycle through the objects and check if they are on the correct side of our hyperplane. If one is not, we make a correction to the hyperplane and continue checking the objects against the new hyperplane. Eventually the hyperplane is adjusted to correctly separate all the objects and we have our decision boundary!
+
+#### A Brief Word on Hyperplanes
+How do you pick your starting hyperplane? It's up to you! Be creative! Or not... Actually perceptron's convergence times are sensitive to conditions such as the initial hyperplane and even the order the objects are looped through, so you might not want to go too wild.
+
+We will be dealing with a two dimensional space, so our divider will be a line. The standard equation for a line is usually written as $y = ax+b$, where $a,b \in \Re$, however, in order to help generalize the idea to higher dimensions, it's convenient to reformulate this equation as $w_0 + w_1x + w_2y = 0$. This is the form of the [hyperplane](https://en.wikipedia.org/wiki/Hyperplane) we will be using, so your output should be $[w_0, w_1, w_2]$. In machine learning, ${w_0,w_1,w_2}$ are usually referred to as weights. 
+
+While hyperplanes are equivalent under scalar multiplication, there is a difference between $[w_0, w_1, w_2]$ and $[-w_0, -w_1, -w_2]$ in that the normal to the hyperplane points in opposite directions. By convention, the perceptron normal points towards the class defined as positive, so this property will be checked but not result in a test failure.
```

Comment:
I think this should maybe be a test failure. It would simplify the messaging for the student in the instruction and from the tests. Do you think that would be okay?

Reply:
Also, negative numbers are scalars (at least to me?), so I would probably drop the first part of the first sentence.

Reply:
>I think this should maybe be a test failure. It would simplify the messaging for the student in the instruction and from the tests. Do you think that would be okay?

I didn't want to strictly penalize a flipped normal since it is just a convention (logical though it may be) and it would at still tell the student they have successfully found a decision boundary. In any case, I think it would be nice to leave this in, but if you think it's better to enforce the convention (which of course is good practice), we can turn this into a failed test, with it's own error message.

>Also, negative numbers are scalars (at least to me?), so I would probably drop the first part of the first sentence.

I've rephrase the scalar multiplication part to make more clear what I was trying to say (that a hyperplane looks the same multiplied by any scalar, but negative scalars flip what we define to be the normal). Hopefully that's clearer :)

Reply:
I think if we ask the student to define a classify(boundary, point) function then they'll have to get consistent about the boundary anyway?

See https://github.com/exercism/julia/pull/678#discussion_r1477341834

Reply:
Edit: moving this down here to split up topics

> I agree that testing with unseen points is a bit tricky, but it's not super hard given that we control the input data and it's not too bad for us to either only include hand-written tests or to do some geometry to ensure our unseen pointd will always be classified correctly by a valid linear boundary.

I was thinking about this, and it appears it is redundant to test unseen points which are guaranteed to be on the correct side of any possible decision boundary for a population. That's because there are two hyperplanes which define the limits of a cone of possibility, and each of these limits can be defined by two points, one from each class. These two points work like support vectors, and are robust to any additional (guaranteed valid) unseen points. This means only the support vectors need to be classified correctly to verify that the hyperplane is a valid decision boundary, and this is already done under the current testing.

For this reason I think it would be best to leave out testing of unseen points, since, for students encountering this for the first time, it could give the unintended impression that Perceptron will always correctly classify unseen points.

Any thoughts?

Reply:
I don't think these objections are relevant?

The tests exist to guide the students to a correct solution. It's fine and expected if the points we provide are guaranteed to be correctly classified *by a correct implementation*. The point of the tests is just to catch bad implementations.

Reply:
Thanks for explaining :)

I want us to have at least one test with unseen points (and ideally more) because I believe that:

1. the tests will make more sense to students who read them
2. confused students won't be able to encode the seen points in the perceptron and pass the tests.

I agree that the tests should be unambiguous. We can have unseen points without producing ambiguous tests by picking points that are geometrically guaranteed to be on the right side of any valid linear boundary between the classes.

E.g. for linearly separable datasets, I think the centroid of any triangle drawn from points of the same set will be classified correctly by any correct implementation.

Another option would be to train on and classify larger datasets and assert that the accuracy is above some threshold that it is either impossible or implausible for a correct implementation to miss.

We don't need to provide the code for generating the test data to the student. We can precompute some datasets and include them in the test file.

I don't think that it's necessary to test the training and classification separately, though we can choose to do so. I think it's fine to just call both functions for every test, as shown above.

Does that make sense?

Reply:
I believe I understand your intentions, but I want to check if I'm getting everything:

> 2. Confused students won't be able to encode the seen points in the perceptron and pass the tests.

I can possibly understand this two ways, so I'll take a stab at both: 

- Do you mean students won't have seen-points to use locally to debug? If so, the first set of fixed tests have the points and labels shown explicitly. 
- Or, do you mean they might hard code the answers into their function in order to pass the tests (i.e. 'cheat')? If so, the pseudorandom tests don't explicitly show the seen points, and someone would have to hard code over 1200 points across 40+ tests to pass, which, while doable, seems unlikely (also couldn't they just look at other students' solutions after submitting once?).

> We can have unseen points without producing ambiguous tests by picking points that are geometrically guaranteed...

This actually still leaves an ambiguity. If a student presents an incorrect decision boundary (DB), the unseen point is not guaranteed to be either correctly or incorrectly classified. It all depends on the particular DB. Furthermore, if the point happens to be classified correctly, it can give the impression that the DB is _correct_ instead of properly relaying the actual information of being just _not incorrect_, thereby misleading the student.

Just a note: the reason I said that any test that can be done with unseen points can be done as well or better with seen-points is that correct classification of a maximum of four special points (support vectors) out of the seen-points is both necessary and sufficient to tell if the DB is 100% correct or not. _Any other test is technically deficient on its own, and deficiency introduces ambiguity_. On the other hand, any other in conjunction with classifying the support vectors is superfluous. However, finding them is a non-trivial task, so every point is exhaustively classified in testing (to make sure they are found and tested).

> Another option would be to train on and classify larger datasets and assert that the accuracy is above some threshold that it is either impossible or implausible for a correct implementation to miss.

I believe this is what the pseudorandom test set does and accuracy is required to be 100%. Is there something further that I'm missing?

> I think it's fine to just call both functions for every test.

I believe this adds unnecessary confusion when there is a failed test as to which function is causing the failure.

> 1. the tests will make more sense to students who read them.

At this point, even if this is the case, the added possibility of confusion created through various ambiguities, and/or the difficulties in trying to mitigate them, make including these features seem to be not worth it. I don't mind including ambiguity in exercises, since it provides 'real world' experience, but it depends on the audience. From what I gather from your concern in this point, you are looking out for less experienced students, so I would presume less ambiguity would be better?

>E.g. for linearly separable datasets, I think the centroid of any triangle drawn from points of the same set will be classified correctly by any correct implementation.

Thanks! I was trying to think of the simple way you had mentioned, but I had lost the trees for the forest :)

Reply:
> I don't think that it's necessary to test the training and classification separately, though we can choose to do so. I think it's fine to just call both functions for every test, as shown above.

Sorry, I just wanted to clarify, I am on board with a classification function if it's tested separately before being used in conjunction with the perceptron function. It's really the unseen points I'm most hesitant to accept.

Reply:
I think your belief that unseen points will cause ambiguity (test failures for valid decision boundaries) is making you think that they're a big problem.

But, as I hope the triangle centroid thing shows, we can generate unseen points with zero chance of them causing false test failures.

If you don't believe the triangle thing, then maybe you can believe that, for separable datasets, any point contained by the convex hull of the points of one label will be given that label by any boundary-based classifier that has zero error on the training set. For a linear boundary and a 2D space that one should be provable geometrically with a ruler and some drawings.

Reply:
I think I have failed to unmuddle my concern :) I'm okay with testing to demonstrate how classification of unseen points works, but testing as proposed is too easily confused with testing unseen points to verify the correctness of a decision boundary. My concerns about why testing unseen points to verify correctness is undesirable follow:

I'm with you on the centroid idea (which I quite liked) and familiar with the other arguments you've given, so I'm not concerned about test failures for _valid_ decision boundaries. The minor issue I have in the case of valid decision boundaries is that testing with unseen points to validate a decision boundary is simply redundant. Why it is redundant can be gleaned from your arguments on why and how we can choose correct unseen points: namely they are chosen entirely in reference to seen points.

One issue that concerns me involves _invalid_ decision boundaries. With an invalid decision boundary, there is no guarantee how an unseen point will be classified (just as there is no guarantee how _any_ point will be classified). There are two cases:

1. The unseen point is classified incorrectly and there is a test failure. The student is alerted to the fact that their decision boundary is incorrect. This I'm okay with, although it is still redundant.
2. The unseen point is classified correctly and the test passes. The student is not alerted to the fact their decision boundary is incorrect and either receives conflicting information from other tests or may indeed end up believing that their decision boundary is correct. This is the first aspect I dislike.

The other issue that concerns me is on a more psychological note: People often take partial information, over-generalize and create false impressions (e.g. stereotypes/prejudices/etc). That's how some of the more naive or inattentive students, seeing testing on unseen points in this way, could walk away with the mistaken impression that Perceptron can correctly classify _any_ unseen point (not just carefully selected points). This is the second aspect I dislike.

So, with what the possible outcomes are from testing unseen points, redundant good info or unnecessary potentially misleading info, and the possible misconception about Perceptron that can be conveyed simply by including the testing, I don't feel the testing of unseen points in the proposed manner brings enough to the table to be included.

**To boil it all down: I feel that 'testing' unseen points to demonstrate how classification works has value, but I think this is too easily confused with testing to verify correctness of the decision boundary. I currently can't think of tests which can eradicate my concerns above.** However, there are related ideas we could demonstrate.

For example, we could show that the introduction of a new point to the data will (in all likelihood) result in a different decision boundary. This doesn't involve classification of unseen points though. As an aside: with SVM, we could introduce an element of a demonstration of unseen point classification switching with a before and after approach, but it just doesn't work with Perceptron because of the non-uniqueness of the decision boundary (i.e. we can't be guaranteed a switch in classification unless we control the initial hyperplane the student uses).

Do you have any other ideas on ways to demonstration classification of unseen points?

Reply:
The cost vs benefit analysis that is going on in my head is the following:

1. Testing unseen points to verify correctness is both redundant and carries downsides
2. Testing unseen points to demonstrate classification is desirable, but unnecessary to satisfy the aim of the exercise.

The exercise is first and foremost meant to have students learn about the Perceptron algorithm. Including a demonstration of how classification works would be a bonus, but I just am not sure this type of testing environment properly supports demonstration of this kind.